pkg/reconciler/pipelinerun/pipelinerun.go

```diff
@@ -605,27 +605,35 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1.PipelineRun, getPipel
 			return controller.NewPermanentError(err)
 		}
 
-		// Make an attempt to create Affinity Assistant if it does not exist
-		// if the Affinity Assistant already exists, handle the possibility of assigned node becoming unschedulable by deleting the pod
-		if !c.isAffinityAssistantDisabled(ctx) {
-			// create Affinity Assistant (StatefulSet) so that taskRun pods that share workspace PVC achieve Node Affinity
-			// TODO(#6740)(WIP): We only support AffinityAssistantPerWorkspace at the point. Implement different AffinityAssitantBehaviors based on `coscheduling` feature flag when adding end-to-end support.
-			if err = c.createOrUpdateAffinityAssistantsPerAABehavior(ctx, pr, affinityassistant.AffinityAssistantPerWorkspace); err != nil {
-				logger.Errorf("Failed to create affinity assistant StatefulSet for PipelineRun %s: %v", pr.Name, err)
-				pr.Status.MarkFailed(ReasonCouldntCreateOrUpdateAffinityAssistantStatefulSetPerWorkspace,
-					"Failed to create StatefulSet for PipelineRun %s/%s correctly: %s",
-					pr.Namespace, pr.Name, err)
-				return controller.NewPermanentError(err)
```

Comment:
At this point, we are creating PVCs from VolumeClaimTemplate before creating AA if it the behavior is `coschedule workspace` or `disabled`. We skip such logic for `coschedule pipelinerun` mode

Reply:
hm I think I understand; possible suggestion to make the code a bit cleaner: what if we move PVC creation to `createOrUpdateAffinityAssistants`? It would help that function make fewer assumptions about what pvcs exist and make handling the disabled case less redundant. Up to you.

Reply:
hmmm, as a developer reading code, I wouldn't expect the pvc creation is in the `createOrUpdateAffinityAssistants` especially when AA is disabled. I feel it could be better to separate concern bettween PVC creation and AA?

Reply:
Synced with @lbernick offline, we agreed to put the PVC creation logic in the Affinity Assistant creation function to make it atomic.